---
title: "Analyzing R-loop data with RLSeq"
author:
  - name: "Henry Miller"
    affiliation: 
      - Alex Bishop Laboratory, UT Health San Antonio
      - Bioinformatics Research Network
package: RLSeq
abstract: |
  This vignette covers basic usage of RLSeq, part of [RLSuite](www.path_to_rlsuite.com), for evaluating data quality and 
  analyzing R-loop locations in comparison with [RLBase](www.rlbase_website.com).
output: 
  BiocStyle::html_document:
    toc_float: true
  BiocStyle::pdf_document: default
vignette: >
  %\VignetteIndexEntry{Analyzing R-loop Data with RLSeq}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}  
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
BiocStyle::markdown()
```

# Introduction

__RLSeq__ is a toolkit for analyzing and evaluating R-loop mapping data sets and it is a core component of the [RLSuite](www.path_to_rlsuite) tool chain. It serves two primary purposes: (1) to facilitate the evaluation of data quality, and (2) to enable R-loop data analysis in the context of genomic annotations and public data sets in [RLBase](www.rlbase_website.com). The package is intended to provide a simple pipeline, called with the `RLSeq()` function, which performs all main analyses and outputs a user-friendly HTML report. Individual functions are also accessible and provide custom analysis capabilities.

This vignette will showcase the primary functionality of RLSeq through an analysis of a publicly-available R-loop data set. It will proceed to discuss, in detail, the specific steps of a typical analysis along with caveats and other considerations. 

# The Standard RLSeq workflow

## Quick-start

The following details a simple analysis workflow that can be run without any prior data, due to the availability of processed data sets in RLBase and annotations provided by RLHub. The commands below download these data sets and load the annotations. Finally, they run `RLSeq()`, which performs all core analysis steps and generates an HTML report. 

```R
# Peaks and coverage can be found in RLBase
rlbase <- "https://rlbase-data.s3.amazonaws.com"
pks <- file.path(rlbase, "peaks", "SRX1025890_hg38.broadPeak")
cvg <- file.path(rlbase, "coverage", "SRX1025890_hg38.bw")

# Annotations can be found in RLHub
annot <- file.path(rlbase, "RLHub", "annotations_hg38.rda")
tmp <- tempfile()
download.file(annot, destfile = tmp)
load(tmp)

# The RLSeq command performs all analysis to generate an HTML report
rls <- RLSeq(
  peaks = pks,
  coverage = cvg,
  annotations = annotations,
  genome = "hg38"
)
```

The report generated by this code is found [here](www.somewhere_on_internet.com).

# RLSeq in full

## Preliminary

### Installation 

RLSeq should be installed alongside [RLHub](www.rlhub_location.com) to facilitate access to the data required for annotation and analysis. 

```R
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("RLSeq")
```

Both packages can be installed from github as well.

```R
library(remotes)
install_github("Bishop-Laboratory/RLHub")
install_github("Bishop-Laboratory/RLSeq")
```

### Obtaining processed R-loop data

RLSeq is compatible with R-loop data generated from a variety of pipelines and tools. _However_, it is strongly recommended that [macs2/macs3](https://github.com/macs3-project/MACS) is used for peak calling as this will improve compatibility with RLBase. In RLSuite, we have provided [RLPipes](www.path_to_rlpipes.com), a snakemake-based CLI pipeline tool built specifically for upstream processing of R-loop datasets. It is also the tool which was used to generate all data within RLBase. 

#### RLPipes for upstream data processing

RLPipes can be downloaded from [conda].

```shell
# conda install -c conda-forge mamba
mamba install -c bioconda rlpipes
```

A typical config file `CSV` file should be written:

|experiment|control   |
|----------|----------|
|SRX113814 |          |
|SRX1025890|SRX1025893|
|SRX1025899|          |

And then the pipeline can be run.

```shell
RLPipes build -m DRIP rseq_out/ tests/test_data/samples.csv
RLPipes run rseq_out/
```

The resulting directory will contain `peaks/`, `coverage/`, `bam/`, and other processed datasets which are directly compatible with RLSeq.

## The RLSeq Analysis Workflow

The following details each step of the analysis pipeline which is run as part of the `RLSeq()` command. We will showcase the usage of each and describe important parameters.

### 

```{r setup}
library(RLSeq)
```




