---
title: "Analyzing R-loop data with RLSeq"
author:
  - name: "Henry Miller"
    affiliation: 
      - Alex Bishop Laboratory, UT Health San Antonio
      - Bioinformatics Research Network
package: RLSeq
abstract: |
  This vignette covers basic usage of RLSeq, part of [RLSuite](www.path_to_rlsuite.com), for evaluating data quality and 
  analyzing R-loop locations in comparison with [RLBase](www.rlbase_website.com).
output: 
  BiocStyle::html_document:
    toc_float: true
  BiocStyle::pdf_document: default
params:
  newdat: 1  # I chose for a reason
vignette: >
  %\VignetteIndexEntry{Analyzing R-loop Data with RLSeq}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}  
bibliography: references.bib
---

```{r, include = FALSE}
knitr::opts_chunk$set(tidy = FALSE,
                      cache = FALSE,
                      dev = "png",
                      message = FALSE, 
                      error = FALSE,
                      warning = FALSE)
BiocStyle::markdown()
```

# Introduction

<img src=`r system.file("Rmd", "logo.png", package = "RLSeq")` align="right" alt="logo" width="240" style = "border: none; float: right;">

**RLSeq** is a toolkit for analyzing R-loop mapping data sets and it is a core component of the *RLSuite* tool chain. It serves two primary purposes: (1) to facilitate the evaluation of data quality, and (2) to enable R-loop data analysis in the context of genomic annotations and the public data sets in [RLBase](www.rlbase_website.com). The package is intended to provide a simple pipeline, called with the `RLSeq()` function, which performs all main analyses. Then, an HTML report can be generated using the `report()` function. Individual functions are also accessible and provide custom analysis capabilities.

This vignette will showcase the primary functionality of RLSeq with a publicly-available R-loop data set which contains a positive (S9.6 -RNaseH1) and negative (S9.6 +RNaseH1) DNA-RNA Immunoprecipitation sequencing (DRIP-seq) sample. It will proceed to discuss, in detail, the specific steps of the typical analysis.

# Quick-start

Here, we demonstrate a simple analysis workflow which utilizes a publicly-available data set stored in *RLBase*. The commands below download these data, run `RLSeq()`, and generate the HTML report.

``` {.r}
# Peaks and coverage can be found in RLBase
rlbase <- "https://rlbase-data.s3.amazonaws.com"
pks <- file.path(rlbase, "peaks", "SRX1025890_hg38.broadPeak")
cvg <- file.path(rlbase, "coverage", "SRX1025890_hg38.bw")

# Initialize data in the RLRanges object. 
# Metadata is optional, but improves the interpretability of results
rlr <- RLRanges(
  peaks = pks,
  coverage = cvg,
  genome = "hg38",
  mode = "DRIP",
  condType = "POS",
  sampleName = "TC32 DRIP-Seq"
)

# The RLSeq command performs all analyses
rlr <- RLSeq(rlr)

# Generate an html report
report(rlr, reportPath = "rlseq_report.html")
```

The report generated by this code is found [here](https://rlbase-data.s3.amazonaws.com/misc/rlseq_report.html).

# End-to-end analysis

## Preliminary

### Installation

RLSeq should be installed alongside [RLHub](https://github.com/Bishop-Laboratory/RLHub) to facilitate access to the data required for annotation and analysis. When downloading RLSeq from bioconductor, RLHub is already included.

``` {.r}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("RLSeq")
```

Both packages can also be installed from github.

``` {.r}
library(remotes)
install_github("Bishop-Laboratory/RLHub")
install_github("Bishop-Laboratory/RLSeq")
```

### Obtaining data

RLSeq is compatible with R-loop data generated from a variety of pipelines and tools. *However*, it is strongly recommended that [macs2/macs3](https://github.com/macs3-project/MACS) is used for peak calling as this will improve compatibility with RLBase. In RLSuite, we have provided [RLPipes](https://github.com/Bishop-Laboratory/RLPipes), a snakemake-based CLI pipeline tool built specifically for upstream processing of R-loop datasets. It is also the tool which was used to generate all data within RLBase.

#### RLPipes for upstream data processing

RLPipes can be installed using [mamba](https://anaconda.org/conda-forge/mamba) or [conda](https://docs.conda.io/en/latest/miniconda.html) (slower).

``` {.shell}
# conda install -c conda-forge mamba
mamba create -n rlpipes -c bioconda rlpipes
```

A typical config file `CSV` file should be written:

| experiment |
|:---------- |
| SRX1025890 |
| SRX1025892 |

And then the pipeline can be run.

``` {.shell}
RLPipes build -m DRIP rseq_out/ tests/test_data/samples.csv
RLPipes run rseq_out/
```

The resulting directory will contain `peaks/`, `coverage/`, `bam/`, and other processed datasets which are directly compatible with RLSeq.

## The RLSeq Workflow

The following details each step of the analysis pipeline which is run as part of the `RLSeq()` command. We will showcase the usage of each and describe important parameters.

```{r library, echo=FALSE}
library(RLSeq)
library(dplyr)
```

### Data sets

For this example, we will be using data from a 2018 *Nature* paper on R-loops in Ewing sarcoma [@Gorthi2018a]. Two samples are used, one which has been IP'd for R-loops (S9.6), and one which is the same, but with the addition of an RNaseH1 treatment (S9.6 + RNaseH1). RNaseH1 degrades RNA:DNA hybrids (a core component of R-loops), so it serves as a useful negative control.

| experiment | condition      |
|:---------- |:-------------- |
| SRX1025890 | S9.6           |
| SRX1025892 | S9.6 + RNaseH1 |

: R-loop mapping samples in Ewing sarcoma cells

The data was processed using RLPipes and uploaded to RLBase. We begin by reading in the data to `GRanges` objects using a helper function from `regioneR`.

```{r}
rlbase <- "https://rlbase-data.s3.amazonaws.com"

# Get peaks and coverage
s96Pks <- regioneR::toGRanges(file.path(rlbase, "peaks", "SRX1025890_hg38.broadPeak"))
s96Cvg <- file.path(rlbase, "coverage", "SRX1025890_hg38.bw")
rnhPks <- regioneR::toGRanges(file.path(rlbase, "peaks", "SRX1025892_hg38.broadPeak"))
rnhCvg <- file.path(rlbase, "coverage", "SRX1025892_hg38.bw")

## Build RLRanges ##

# S9.6 -RNaseH1
rlr <- RLRanges(
  peaks = s96Pks, 
  coverage = s96Cvg,
  genome = "hg38",
  mode = "DRIP",
  condType = "POS",
  sampleName = "TC32 DRIP-Seq",
  quiet = TRUE
)

# S9.6 +RNaseH1
rlrRNH <- RLRanges(
  peaks = rnhPks, 
  coverage = rnhCvg,
  genome = "hg38",
  mode = "DRIP",
  condType = "NEG",
  sampleName = "TC32 DRIP-Seq (+RNaseH1)",
  quiet = TRUE
)
```

### Sample quality

Sample quality is assessed by analyzing the association of peaks with R-loop-forming sequences (RLFS). RLFS are genomic sequences that favor the formation of R-loops [@Jenjaroenpun2015]. While R-loops can form outside RLFS, there is a noticeable relationship between R-loops and RLFS, which provides an unbiased test of whether a set of peaks actually represents successful R-loop mapping. This method is described in full within the RLSuite manuscript.

#### Permutation tests

RLSeq first implements a permutation test to evaluate the enrichment of peaks within RLFS and build a Z-score distribution around the TSS.

```{r rlfs-test}
# Analyze RLFS for positive sample
rlr <- analyzeRLFS(rlr, quiet = TRUE)
```

The resulting objects now contain the permutation test results. These results can be easily visualized with the `permPlot` function.

```{r plot-perm, fig.cap="Plot of permutation test results (S9.6)."}
plotRLFSRes(rlr)
```

As a comparison, we also test the negative control (RNaseH1-treated) samples.

```{r plot-perm-rnh, fig.cap="Plot of permutation test results (S9.6 + RNaseH1)."}
rlrRNH <- analyzeRLFS(rlrRNH, quiet = TRUE)
plotRLFSRes(rlrRNH)
```

From this example, we can see a fundamental challenge facing RLFS-based quality assessment: while the RNaseH1-treated sample is clearly a negative control, and the distribution does not resemble examples of positive samples found in RLBase [link](https://rlbase-data.s3.amazonaws.com/misc/examples_for_select_samples.png), the p-value is still significant at `p < 0.0099`. This indicates that permutation testing alone is insufficient to distinguish successful and unsuccessful R-loop mapping. This observation motivated the development of a classifier model, which was then developed and provided for use with RLSeq.

#### Classifier predictions

The classifier is an ensemble model based on an online-learning scheme as detailed in the RLSuite manuscript. The latest version can be accessed via [RLHub](https://github.com/Bishop-Laboratory/RLHub).

```{r predict-condition}

# Predict 
rlr <- predictCondition(rlr)
rlrRNH <- predictCondition(rlrRNH)

# Access results
s96_pred <- rlresult(rlr, "predictRes")
rnh_pred <- rlresult(rlrRNH, "predictRes")

# Results
tibble::tibble(
  condition = c("S9.6", "S9.6 + RNaseH1"),
  verdict = c(s96_pred$Verdict, rnh_pred$Verdict)
) %>% 
  kableExtra::kable(format = "html") %>% 
  kableExtra::kable_classic(full_width=FALSE, font_size = 20)
```

### Feature enrichment

Now that we have verified the quality of our S9.6 sample (SRX1025890), we proceed to perform a feature enrichment test. This allows us to determine what genomic features were enriched within the mapped R-loops from this data set. The test involves querying the annotation database in RLHub and performing fisher's exact test as well as the relative distance test [@Favorov2012].

```{r}

# Annotations can be found in RLHub
annot <- file.path(rlbase, "RLHub", "annotations_primary_hg38.rda")
tmp <- tempfile()
download.file(annot, destfile = tmp, quiet = TRUE)
load(tmp)

# Perform test
rlr <- featureEnrich(
  object = rlr,
  annotations = annotations_primary,
  quiet = TRUE
)

# View Top Results
annoResS96 <- rlresult(rlr, "featureEnrichment")
annoResS96 %>%
  relocate(contains("fisher"), .after = type) %>%
  arrange(desc(stat_fisher_rl)) %>%
  slice_head(n = 10) %>%
  DT::datatable(options = list(scrollX=TRUE))
```

From the top results (ranked by Fisher's exact test odds ratio), we observe a high degree of enrichment within genic features, such as exons and introns, as well as regulatory features such as enhancer promoters (enhP). These results are consistent with previous reports of R-loop location mapping [@Castillo-Guzman2021].

<details>
<summary>RNaseH1 Comparison</summary>

As a comparison, we also performed the same test with the RNaseH1-treated sample.

```{r}
# Perform test
rlrRNH <- featureEnrich(
  object = rlrRNH,
  annotations = annotations_primary,
  quiet = TRUE
)

# View Top Results
annoResRNH <- rlresult(rlrRNH, "featureEnrichment")
annoResRNH %>%
  relocate(contains("fisher"), .after = type) %>%
  arrange(desc(stat_fisher_rl)) %>%
  DT::datatable()
```

We found that the top hits were different from those uncovered in the positive (S9.6 -RNaseH1) condition. Particularly, there was stark enrichment for repetitive elements and a diminished enrichment for genomic features.

To visualize these results, we pull the top 6 enriched annotations within experiment and control samples and plotted them as a heat map.

```{r}
inner_join(
  annoResRNH, 
  annoResS96,
  by = c("db", "type"),
  suffix = c("__S9.6 +RNaseH1", "__S9.6 -RNaseH1")
) %>%
  select(db, type,  contains("stat_fisher_rl")) %>%
  tidyr::pivot_longer(cols = contains("__")) %>%
  mutate(
    group = gsub(name, pattern = ".+__(.+)$", replacement = "\\1"),
    value = log2(value)
  ) %>%
  filter(! is.na(value),
         is.finite(value)) %>%
  group_by(group) %>%
  filter(type %in% (slice_max(., order_by = value, n = 8) %>% pull(type))) %>%
  ungroup() %>%
  select(value, group, type) %>%
  tidyr::pivot_wider(id_cols = type, names_from = group, 
                     values_from = value, values_fill = 0) %>%
  tibble::column_to_rownames(var = "type") %>%
  as.matrix() %>%
  ComplexHeatmap::pheatmap(
    color = colorRampPalette(RColorBrewer::brewer.pal("PuBuGn", n = 9))(100), 
    main = "Top enriched features",
    name = "Fisher Log2 Ratio", 
    angle_col = "45"
  )
```

From this result, we plainly observe the stark contrast in enrichment between samples which map R-loops (S9.6 -RNaseH1) and those which do not (S9.6 + RNaseH1).

<br>

</details>

<br>

#### Visualization of enrichment results

RLSeq provides a helper function, `plotEnrichment`, to facilitate the visualization of enrichment results.

```{r, warning=FALSE, figures-side, fig.show="hold", out.width="50%"}
pltlst <- plotEnrichment(rlr)
```

Encode cis-regulatory elements (CREs):

```{r warning=FALSE}
pltlst$Encode_CREs
```

Known Gene RNAs:

```{r warning=FALSE}
pltlst$knownGene_RNAs
```

<details>
<summary>Using the `splitby` parameter</summary>

The `splitby` parameter allows for plotting with respect to `verdict` or `condType`. This can be useful when trying to determine whether a particular annotation is uniquely-enriched in high-quality samples. We also set `onlyCase = FALSE` so that both "Case" and "Control" samples are displayed.

```{r}
pltlst <- plotEnrichment(rlr, splitby = "verdict", onlyCase = FALSE)
```

Encode cis-regulatory elements (CREs):

```{r}
pltlst$Encode_CREs
```

</details>

### Correlation analysis

Correlation analysis finds inter-sample correlation coefficients by using bin-level R-loop counts around gold-standard R-loop sites (sites profiled using ultra-long-read R-loop mapping -- "SMRF-Seq") [@Chédin2021].

```{r}
rlr <- corrAnalyze(rlr)
```

The results of this analysis may be visualized using `corrHeatmap`.

```{r, fig.height=5, fig.width=6.8}
corrHeatmap(rlr)
```

### Gene Annotation

RLSeq provides gene annotation of RLRanges. Briefly, annotations are automatically downloaded using `AnnotationHub()` and then intersected with RLRanges.

```{r}
rlr <- geneAnnotation(rlr, quiet=TRUE)
```

<details>
<summary>Over-representation analysis</summary>

These results can then be used for over-representation analysis if desired. One caveat to this analysis is that the number of genes is a function of the number of peaks, and this can lead (at higher numbers) to noticeable bias in enrichment analysis.

```{r}
len <- rlresult(rlr, "geneAnnoRes") %>%
  pull(gene_id) %>%
  unique() %>%
  length()
cat(len, " genes found in RLRanges")
```

To address this, we filter RLRanges object to obtain the top 2000 ranges by p-adjusted value. Because our ranges are "broadPeak" (see specification [here](https://genome.ucsc.edu/FAQ/FAQformat.html#format13)), the 9th column contains these values.

```{r}
# Pull the peak names for the top 2000 peaks
topPks <- rlr %>%
  as.data.frame() %>%
  tibble::rownames_to_column(var = "peakName") %>%
  slice_max(V9, n = 2000) %>%
  pull(peakName)

# Filter the results to obtain the corresponding genes
rlgenes <- rlresult(rlr, "geneAnnoRes") %>%
  filter(peak_name %in% {{ topPks }}) %>%
  pull(gene_id) %>%
  unique()

# Get lenght of these genes
cat(length(rlgenes), " genes found in top 2000 RLRanges")
```

Before we can perform pathway enrichment, we convert our gene IDs to gene symbols.

```{r}
symbols <- AnnotationDbi::mapIds(
  org.Hs.eg.db::org.Hs.eg.db, 
  keys = rlgenes, 
  keytype = "ENTREZID", 
  column = "SYMBOL"
)
```

With our final list of genes prepared, we proceed to perform over-representation analysis using the [enrichr](https://maayanlab.cloud/Enrichr/) web service.

```{r}
response <- httr::POST(
  url = 'https://maayanlab.cloud/Enrichr/addList', 
  body = list(
    'list' = paste0(symbols, collapse = "\n"),
    'description' = paste0("RL-overlap Genes from ", rlr@metadata$sampleName)
  )
)
response <- jsonlite::fromJSON(httr::content(response, as = "text"))  
permalink <- paste0(
  "https://maayanlab.cloud/Enrichr/enrich?dataset=", response$shortId[1]
)
```

The permalink to these results can be found <a href="`r permalink`" target="_blank">here</a>.

</details>

### R-Loop Region Test

R-loop regions are consensus sites for R-loop formation discovered from profiling of \> 200 high-confidence R-loop mapping samples in RLBase. The degree of overlap between any given sample and RL-regions is a useful quality metric which has the added benefit of providing additional context on the type of R-loops which have been profiled.

```{r}
rlr <- rlRegionTest(rlr)
```

The test results can be easily visualized in the following manner.

```{r}
plotRLRegionOverlap(rlr)
```

# Session

<details>
<summary>Session info</summary>

```{r}
sessionInfo()
```

</details>


# References
