### Mapping quality

```{r mapping quality}
reads_aligned <- bam_stats$reads_aligned / 1e6
duplicate_reads <- bam_stats$duplicate_reads / 1e6

if (show_readqc) {
  total_reads <- bam_stats$total_reads / 1e6
  total_reads_color <- "black"
  pct_aligned <- round(100*(reads_aligned/total_reads), digits = 2)
  pct_align_color <- ifelse(pct_aligned > 80, "green", ifelse(pct_aligned > 60, "orange", "red"))
  pct_duplicated <- round(100*(duplicate_reads/total_reads), digits = 2)
  pct_duplicated_color <- ifelse(pct_duplicated < 35, "green", ifelse(pct_aligned < 50, "orange", "red"))
} else {
  total_reads <- "N/A"
  total_reads_color <- "grey"
  pct_aligned <- "N/A"
  pct_align_color <- "grey"
  pct_duplicated <- "N/A"
  pct_duplicated_color <- "grey"
}

genome <- configlist$genome
```

Reads were aligned to **`r genome`** with `bwa mem`. 

Total reads (millions): <strong style="color: `r total_reads_color`">`r total_reads`</strong>

Aligned reads (millions): **`r reads_aligned`**

Percent of reads aligned: <strong style="color: `r pct_align_color`">`r pct_aligned`</strong>

Duplicated reads (millions): **`r duplicate_reads`**

Percent of reads duplicated: <strong style="color: `r pct_duplicated_color`">`r pct_duplicated`</strong>

***
