<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Available Genomes — available_genomes • RLSeq</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Available Genomes — available_genomes" />
<meta property="og:description" content="Contains metadata about all the genomes available in UCSC. It contains
derived metadata, such as the effective genome sizes as well. See also
the data-raw/available_genomes.R script to see processing steps." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RLSeq</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.9</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/RLSeq.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/Bishop-Laboratory/RLSeq/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Available Genomes</h1>
    <small class="dont-index">Source: <a href='https://github.com/Bishop-Laboratory/RLSeq/blob/master/R/data.R'><code>R/data.R</code></a></small>
    <div class="hidden name"><code>available_genomes.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Contains metadata about all the genomes available in UCSC. It contains
derived metadata, such as the effective genome sizes as well. See also
the <code>data-raw/available_genomes.R</code> script to see processing steps.</p>
    </div>

    <pre class="usage"><span class='va'>available_genomes</span></pre>


    <h2 class="hasAnchor" id="format"><a class="anchor" href="#format"></a>Format</h2>

    <p>An object of class <code>data.frame</code> with 199 rows and 27 columns.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    
<h3 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Structure</h3>


<p><code>available_genomes</code> is a <code>data.frame</code> with the following columns:</p><ul>
<li><p><code>UCSC_orgID</code></p><ul>
<li><p>Official UCSC ID of the genome</p></li>
</ul></li>
<li><p><code>description</code></p><ul>
<li><p>Verbose description of the assembly, source, and year/month of entry.</p></li>
</ul></li>
<li><p><code>nibPath</code></p><ul>
<li><p>Endpoint of the genome in UCSC gbdb.</p></li>
</ul></li>
<li><p><code>organism</code></p><ul>
<li><p>Name of the organism.</p></li>
</ul></li>
<li><p><code>defaultPos</code></p><ul>
<li><p>Default location of genome browser view for this genome.</p></li>
</ul></li>
<li><p><code>active</code></p><ul>
<li><p>Description not available.</p></li>
</ul></li>
<li><p><code>orderKey</code></p><ul>
<li><p>Description not available.</p></li>
</ul></li>
<li><p><code>genome</code></p><ul>
<li><p>The name of the genome.</p></li>
</ul></li>
<li><p><code>scientificName</code></p><ul>
<li><p>The scientific name of the organism.</p></li>
</ul></li>
<li><p><code>htmlPath</code></p><ul>
<li><p>Path in UCSC gbdb to the <code>description.html</code> file for the genome.</p></li>
</ul></li>
<li><p><code>hgNearOk</code></p><ul>
<li><p>Description not available.</p></li>
</ul></li>
<li><p><code>hgPbOk</code></p><ul>
<li><p>Description not available.</p></li>
</ul></li>
<li><p><code>sourceName</code></p><ul>
<li><p>Name of organization providing the genome.</p></li>
</ul></li>
<li><p><code>taxId</code></p><ul>
<li><p>The taxonomy ID of the organism.</p></li>
</ul></li>
<li><p><code>genes_available</code></p><ul>
<li><p>If <code>TRUE</code>, the gene annotations are available in GTF format.</p></li>
</ul></li>
<li><p><code>year</code></p><ul>
<li><p>The year the genome assembly was added.</p></li>
</ul></li>
<li><p><code>eff_genome_size_XXbp</code></p><ul>
<li><p>The effective genome size of this genome. Calculated at various read
lengths with <a href='https://khmer.readthedocs.io/en/latest/'>khmer</a>
and used to improve the accuracy of analysis. See
the <code>data-raw/available_genomes.R</code> script to see how this calculation
was performed.</p></li>
</ul></li>
<li><p><code>genome_length</code></p><ul>
<li><p>The total length of the genome.</p></li>
</ul></li>
<li><p><code>rlfs_available</code></p><ul>
<li><p>If <code>TRUE</code>, R-loop forming sequences annotations are available in the
RLBase AWS S3 repository.</p></li>
</ul></li>
</ul>



    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><span class='va'>available_genomes</span>

</pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Henry Miller.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


