<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>RLSeq — RLSeq • RLSeq</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="RLSeq — RLSeq"><meta property="og:description" content="Executes the RLSeq analysis workflow."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RLSeq</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.3.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/RLSeq.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/Bishop-Laboratory/RLSeq/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>RLSeq</h1>
    <small class="dont-index">Source: <a href="https://github.com/Bishop-Laboratory/RLSeq/blob/HEAD/R/RLSeq.R" class="external-link"><code>R/RLSeq.R</code></a></small>
    <div class="hidden name"><code>RLSeq.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Executes the RLSeq analysis workflow.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">RLSeq</span><span class="op">(</span><span class="va">object</span>, quiet <span class="op">=</span> <span class="cn">FALSE</span>, skip <span class="op">=</span> <span class="cn">NULL</span>, <span class="va">...</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>object</dt>
<dd><p>An <a href="RLRanges.html">RLRanges</a> object.</p></dd>


<dt>quiet</dt>
<dd><p>If <code>TRUE</code>, messages are suppressed. Default: <code>FALSE</code>.</p></dd>


<dt>skip</dt>
<dd><p>Analysis steps to skip.
Default: <code>NULL</code>. See <em>details</em> for options.</p></dd>


<dt>...</dt>
<dd><p>Arguments passed to <a href="analyzeRLFS.html">analyzeRLFS</a>.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>An <a href="RLRanges.html">RLRanges</a> object with results available (see <a href="RLResults.html">rlresult</a>).</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>The <code>RLSeq()</code> function does all of the following by default:</p><ol><li><p><strong>RLFS Perm Test</strong>. Runs the <a href="analyzeRLFS.html">analyzeRLFS</a> function to test the
enrichment of user-supplied ranges within R-loop-forming sequences.
<em>Cannot be skipped.</em></p></li>
<li><p><strong>Predict Condition</strong>. Runs the <a href="predictCondition.html">predictCondition</a> function to
predict whether the user-supplied sample robustly maps R-loops or not.
<em>Cannot be skipped.</em></p></li>
<li><p><strong>Noise analysis</strong>. Runs the <a href="noiseAnalyze.html">noiseAnalyze</a> function to analyze the
signal-noise distribution within the supplied dataset. Skip with
<code>skip="noiseAnalyze"</code>.</p></li>
<li><p><strong>Feature enrichment test</strong>. Runs the <a href="featureEnrich.html">featureEnrich</a> function to
test the enrichment of user-supplied ranges within R-loop-relevant
genomic features. Skip with <code>skip="featureEnrich"</code>.</p></li>
<li><p><strong>Transcript Feature Overlap</strong>. Runs the <a href="txFeatureOverlap.html">txFeatureOverlap</a> function to
get the overlap of transcript features and user-supplied peaks.</p></li>
<li><p><strong>Correlation Analysis</strong>. Runs the <a href="corrAnalyze.html">corrAnalyze</a> function to test
the correlation of user-supplied R-loop signal with other samples in
RLBase around "gold-standard" R-loop regions.
Skip with <code>skip="corrAnalyze"</code>.</p></li>
<li><p><strong>Gene annotation</strong>. Runs the <a href="geneAnnotation.html">geneAnnotation</a> function to find overlap
of genes with the user-supplied ranges. Skip with <code>skip="geneAnnotation"</code>.</p></li>
<li><p><strong>R-loop Region Analysis</strong>. Runs the <a href="rlRegionTest.html">rlRegionTest</a> function to find
the overlap of user-supplied ranges with consensus R-loop sites
(RL-Regions). Skip with <code>skip="rlRegionTest"</code>.</p></li>
</ol></div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Example RLRanges</span></span></span>
<span class="r-in"><span><span class="va">rlr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"rlrsmall.rds"</span>, package <span class="op">=</span> <span class="st">"RLSeq"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Run RLSeq</span></span></span>
<span class="r-in"><span><span class="co"># `useMask=FALSE`, `ntime=10`, and `skip=` for demonstration purposes here.</span></span></span>
<span class="r-in"><span><span class="va">rlr</span> <span class="op">&lt;-</span> <span class="fu">RLSeq</span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="va">rlr</span>,</span></span>
<span class="r-in"><span>    useMask <span class="op">=</span> <span class="cn">FALSE</span>, ntimes <span class="op">=</span> <span class="fl">10</span>,</span></span>
<span class="r-in"><span>    skip <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>        <span class="st">"featureEnrich"</span>, <span class="st">"corrAnalyze"</span>, <span class="st">"geneAnnotation"</span>, <span class="st">"rlRegionTest"</span></span></span>
<span class="r-in"><span>    <span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [1/8] RLFS Perm Test</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Note: The minimum p-value with only 10 permutations is 0.0909090909090909. You should consider increasing the number of permutations."</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   chr22_KI270733v1_random. Note that ranges located on a sequence whose</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   length is unknown (NA) or on a circular sequence are not considered</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   out-of-bound (use seqlengths() and isCircular() to get the lengths and</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   circularity flags of the underlying sequences). You can use trim() to</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   trim these ranges. See ?`trim,GenomicRanges-method` for more</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   chr22_KI270733v1_random. Note that ranges located on a sequence whose</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   length is unknown (NA) or on a circular sequence are not considered</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   out-of-bound (use seqlengths() and isCircular() to get the lengths and</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   circularity flags of the underlying sequences). You can use trim() to</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   trim these ranges. See ?`trim,GenomicRanges-method` for more</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   chr22_KI270733v1_random. Note that ranges located on a sequence whose</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   length is unknown (NA) or on a circular sequence are not considered</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   out-of-bound (use seqlengths() and isCircular() to get the lengths and</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   circularity flags of the underlying sequences). You can use trim() to</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   trim these ranges. See ?`trim,GenomicRanges-method` for more</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   chr22_KI270733v1_random. Note that ranges located on a sequence whose</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   length is unknown (NA) or on a circular sequence are not considered</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   out-of-bound (use seqlengths() and isCircular() to get the lengths and</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   circularity flags of the underlying sequences). You can use trim() to</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   trim these ranges. See ?`trim,GenomicRanges-method` for more</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   chr22_KI270733v1_random. Note that ranges located on a sequence whose</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   length is unknown (NA) or on a circular sequence are not considered</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   out-of-bound (use seqlengths() and isCircular() to get the lengths and</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   circularity flags of the underlying sequences). You can use trim() to</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   trim these ranges. See ?`trim,GenomicRanges-method` for more</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   chr22_KI270733v1_random. Note that ranges located on a sequence whose</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   length is unknown (NA) or on a circular sequence are not considered</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   out-of-bound (use seqlengths() and isCircular() to get the lengths and</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   circularity flags of the underlying sequences). You can use trim() to</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   trim these ranges. See ?`trim,GenomicRanges-method` for more</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   chr22_KI270733v1_random. Note that ranges located on a sequence whose</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   length is unknown (NA) or on a circular sequence are not considered</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   out-of-bound (use seqlengths() and isCircular() to get the lengths and</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   circularity flags of the underlying sequences). You can use trim() to</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   trim these ranges. See ?`trim,GenomicRanges-method` for more</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   chr22_KI270733v1_random. Note that ranges located on a sequence whose</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   length is unknown (NA) or on a circular sequence are not considered</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   out-of-bound (use seqlengths() and isCircular() to get the lengths and</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   circularity flags of the underlying sequences). You can use trim() to</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   trim these ranges. See ?`trim,GenomicRanges-method` for more</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   chr22_KI270733v1_random. Note that ranges located on a sequence whose</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   length is unknown (NA) or on a circular sequence are not considered</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   out-of-bound (use seqlengths() and isCircular() to get the lengths and</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   circularity flags of the underlying sequences). You can use trim() to</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   trim these ranges. See ?`trim,GenomicRanges-method` for more</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   chr22_KI270733v1_random. Note that ranges located on a sequence whose</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   length is unknown (NA) or on a circular sequence are not considered</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   out-of-bound (use seqlengths() and isCircular() to get the lengths and</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   circularity flags of the underlying sequences). You can use trim() to</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   trim these ranges. See ?`trim,GenomicRanges-method` for more</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   chr22_KI270733v1_random. Note that ranges located on a sequence whose</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   length is unknown (NA) or on a circular sequence are not considered</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   out-of-bound (use seqlengths() and isCircular() to get the lengths and</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   circularity flags of the underlying sequences). You can use trim() to</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   trim these ranges. See ?`trim,GenomicRanges-method` for more</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   chr22_KI270733v1_random. Note that ranges located on a sequence whose</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   length is unknown (NA) or on a circular sequence are not considered</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   out-of-bound (use seqlengths() and isCircular() to get the lengths and</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   circularity flags of the underlying sequences). You can use trim() to</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   trim these ranges. See ?`trim,GenomicRanges-method` for more</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   chr22_KI270733v1_random. Note that ranges located on a sequence whose</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   length is unknown (NA) or on a circular sequence are not considered</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   out-of-bound (use seqlengths() and isCircular() to get the lengths and</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   circularity flags of the underlying sequences). You can use trim() to</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   trim these ranges. See ?`trim,GenomicRanges-method` for more</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   chr22_KI270733v1_random. Note that ranges located on a sequence whose</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   length is unknown (NA) or on a circular sequence are not considered</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   out-of-bound (use seqlengths() and isCircular() to get the lengths and</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   circularity flags of the underlying sequences). You can use trim() to</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   trim these ranges. See ?`trim,GenomicRanges-method` for more</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   chr22_KI270733v1_random. Note that ranges located on a sequence whose</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   length is unknown (NA) or on a circular sequence are not considered</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   out-of-bound (use seqlengths() and isCircular() to get the lengths and</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   circularity flags of the underlying sequences). You can use trim() to</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   trim these ranges. See ?`trim,GenomicRanges-method` for more</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   chr22_KI270733v1_random. Note that ranges located on a sequence whose</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   length is unknown (NA) or on a circular sequence are not considered</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   out-of-bound (use seqlengths() and isCircular() to get the lengths and</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   circularity flags of the underlying sequences). You can use trim() to</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   trim these ranges. See ?`trim,GenomicRanges-method` for more</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   chr22_KI270733v1_random. Note that ranges located on a sequence whose</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   length is unknown (NA) or on a circular sequence are not considered</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   out-of-bound (use seqlengths() and isCircular() to get the lengths and</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   circularity flags of the underlying sequences). You can use trim() to</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   trim these ranges. See ?`trim,GenomicRanges-method` for more</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   chr22_KI270733v1_random. Note that ranges located on a sequence whose</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   length is unknown (NA) or on a circular sequence are not considered</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   out-of-bound (use seqlengths() and isCircular() to get the lengths and</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   circularity flags of the underlying sequences). You can use trim() to</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   trim these ranges. See ?`trim,GenomicRanges-method` for more</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   information.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [2/8] Predict Condition</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> see ?RLHub and browseVignettes('RLHub') for documentation</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> see ?RLHub and browseVignettes('RLHub') for documentation</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [3/8] Noise Analysis</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [4/8] Feature Enrichment Test</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [5/8] Transcript Feature Overlap Analysis</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> see ?RLHub and browseVignettes('RLHub') for documentation</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [6/8] Correlation Analysis</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> No coverage provided... skipping.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [7/8] Gene Annotation</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [8/8] R-loop Region Analysis</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Henry Miller.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

